docs/rfcs/021-raii-runtime.md

```diff
@@ -0,0 +1,209 @@
+# Improving runtime by making use of the RAII idiom
+
+## Motivation
+
+Currently, we have several problems regarding the runtime:
+
+- `TenantState` doesn't clearly show what happens with the tenant. Repository and code comments need to describe how state transitions occur and what they mean. It is critical since the code is going to have more states. The same applies to `TimelineState`.
+- Current tenant and timeline access, shutdown design fails to shut down tenants properly, timelines, background tasks, and control their lifetimes, resulting in races.
+- `tasks_mgr` runtime is redundant, and its design allows us to have even more races without noticing.
```

Comment:
I'm not sure what you mean by redundant. Agreed that it has races, and we need to fix or replace that somehow.

Reply:
Added the description of what means "redundant". IMO the idea of selecting tasks from some global list is just wrong, and it leads to races. The idea is to wait for task completion and store `JoinHandle` directly in the tenant responsible for this task. This is also addressed below.